### PR TITLE
[UI] Implement passwords screen

### PR DESCRIPTION
**Pull request by nicolaslee2775**
connect #348 

Todo: handle template query and mutation #381

##### Password Policy
image
##### Forgot Password
image
##### Warn user when switching tab with unsaved changes
image

### CODE REVIEW

**Comment by nicolaslee2775** [comment]
@louischan-oursky Fixed CI error.

File: portal/src/SwitchTabBlockerDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+  IButtonProps,
+} from "@fluentui/react";
+import { Context, FormattedMessage } from "@oursky/react-messageformat";
+
+export interface SwitchTabBlockerDialogProps extends IDialogProps {
+  onDialogConfirm?: IButtonProps["onClick"];
+  onDialogDismiss?: IButtonProps["onClick"];
+}
+
+const SwitchTabBlockerDialog: React.FC<SwitchTabBlockerDialogProps> = function SwitchTabBlockerDialog(props) {
```
Comment (louischan-oursky) [inline portal/src/SwitchTabBlockerDialog.tsx]: This component does not use any special hooks. NavigationBlockerDialog is special hooks + UI. Can we refactor NavigationBlockerDialog to use SwitchTabBlockerDialog in its implementation? Then this component can simply be renamed to BlockerDialog.

File: portal/src/graphql/portal/PasswordsScreen.tsx [hunk 2]
```diff
@@ -36,6 +42,61 @@ const PasswordsScreen: React.FC = function PasswordsScreen() {
         };
   }, [data]);
 
+  const [currentPivotKey, setCurrentPivotKey] = useState<PivotKey>(PASSWORD_POLICY_PIVOT_KEY);
+  const selectedPivotKeyRef = useRef<PivotKey>(PASSWORD_POLICY_PIVOT_KEY);
```
Comment (louischan-oursky) [inline portal/src/graphql/portal/PasswordsScreen.tsx:56]: We are treating the Pivot as a controlled component. Why do we need to keep ref? If we do not setCurrentPivotKey in onLinkClick, the pivot should not change at all, correct? The logic here is a bit long and complicated.

File: portal/package.json [hunk 3]
```diff
@@ -36,7 +36,7 @@
     "@apollo/client": "3.1.5",
     "@authgear/web": "0.1.0",
     "@fluentui/react": "7.137.1",
-    "@monaco-editor/react": "^3.6.3",
+    "@monaco-editor/react": "^3.5.5",
```
Comment (louischan-oursky) [inline portal/package.json:39]: Why problems does state-local is causing that prevent us from using the latest version of monaco?

Also, use exact version, remove the caret `^`

File: portal/src/graphql/portal/PasswordsScreen.tsx (hunk 2 above)
Comment (nicolaslee2775) [reply portal/src/graphql/portal/PasswordsScreen.tsx:56]: @louischan-oursky We need to switch to other tab when user click confirm button. So need to store the selected key.

File: portal/package.json (hunk 3 above)
Comment (nicolaslee2775) [reply portal/package.json:39]:
```
 ~/Documents/projects/authgear-server/portal  ➦ 4c6b5a57 ●  npm install @monaco-editor/react@3.6.3
 ~/Documents/projects/authgear-server/portal  ➦ 4c6b5a57 ●  npm run build

> authgear-portal@0.1.0 build /Users/nicolaslee/Documents/projects/authgear-server/portal
> PARCEL_BUNDLE_ANALYZER=1 parcel build --no-cache ./src/index.html

🚨 Build failed.
@parcel/core: Failed to resolve 'state-local' from './node_modules/@monaco-editor/react/lib/utils/monaco.js'
/Users/nicolaslee/Documents/projects/authgear-server/portal/node_modules/@monaco-editor/react/lib/utils/monaco.js:14:27
  13 |
> 14 | var _stateLocal = require("state-local");
>    |                           ^^^^^^^^^^^^^
  15 |
  16 | var _config = _interopRequireDefault(require("../config"));
@parcel/resolver-default: Could not load './src/index.js' from module 'state-local' found in package.json#module
/Users/nicolaslee/Documents/projects/authgear-server/portal/node_modules/state-local/package.json:65:13
  64 |   "main": "dist/state-local.js",
> 65 |   "module": "src/index.js",
>    |             ^^^^^^^^^^^^^^ './src/index.js' does not existundefined'
  66 |   "name": "state-local",
  67 |   "repository": {
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! authgear-portal@0.1.0 build: `PARCEL_BUNDLE_ANALYZER=1 parcel build --no-cache ./src/index.html`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the authgear-portal@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/nicolaslee/.npm/_logs/2020-09-24T08_55_23_140Z-debug.log
```

Comment (louischan-oursky) [reply portal/package.json:39]: OK. Let's stick with 3.5.5 then.

File: portal/src/CodeEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    onChange,
+  } = props;
+
+  const callback = useCallback(
+    (e: unknown, value: string | undefined) => {
+      onChange(e, value);
+      return undefined;
+    },
+    [onChange]
+  );
```
[the remark this answers is not in the document]
Comment (nicolaslee2775) [reply portal/src/CodeEditor.tsx]: Updated. Directly passed props to ControlledEditor.

File: portal/src/CodeEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import React, { useCallback } from "react";
+import { ControlledEditor } from "@monaco-editor/react";
+
+export interface CodeEditorOnChange {
+  (e: unknown, value: string | undefined): void;
+}
```
[the remark this answers is not in the document]
Comment (nicolaslee2775) [reply portal/src/CodeEditor.tsx]: Updated. Directly passed props to ControlledEditor.

File: portal/src/SwitchTabBlockerDialog.tsx (hunk 1 above)
Comment (nicolaslee2775) [reply portal/src/SwitchTabBlockerDialog.tsx]: Refactored.

**Comment by louischan-oursky** [comment]
1. Password Strength Level should have a "Off" option, in additional to 1 - 5.
2. The initial state of "Prevent Password Reuse" should be
  - True if either one of the value is non-zero
  - False otherwise
3. See if we can disable axe on monaco

**Comment by axolotl0212** [comment]
I briefly tried out the password screen, I found 2 issues:

1. in number text field, when I input one `.`, it is allowed, when I enter second `.`, input is cleared. Suggest just intercept onChange and match with `/^[0-9]*$/`
2. Monaco edit is extremely slow in Chrome, the page get stuck and unusable
  - could be related to axe scanning DOM tree?

**Comment by nicolaslee2775** [comment]
> 2. Monaco edit is extremely slow in Chrome, the page get stuck and unusable
> 
> * could be related to axe scanning DOM tree?

Yes. I have tried passing `{ exclude: [[".monaco-editor"]] }` to axe.run context but it still super slow.

**Pull request merged**